### DISCUSSION TITLE
Using a templating engine or static site generator to generate HTML?

### DISCUSSION THREAD

**Discussion by prehensileBBC** [obsidian-html/obsidian-html#616] [Ideas]
I've been using obsidian-html to generate a web version of an Obsidian vault I'm using to organise a research project.

I've been tweaking the templates to add features / mess with the layout of the generated HTML.  I've found this somewhat fiddly to do, since the code and templates which contribute to the page building process are scattered across a few different files in the obsidian-html codebase.

I've been wondering if it'd be possible to use a templating engine or static site generator - for example, Jinja2 or Pelican - to handle the HTML building step of generating an HTML site with obsidian-html. This would make editing templates more straightforward, and if Pelican was used, would give access to a lot more themes.

I'd be happy to spend a bit of time looking into this and attempt to implement it - I'll probably give it a try for my own project anyway, but I could also submit a PR if it'd be welcome here too.

**Comment by dwrolvink** [comment]
I only have experience with jinja2, but wouldn't we then still have the issue of having different files for different components? Or would you make one big template and toggle elements on/off based on the config?

Not to say it isn't fiddly or a mess now, but the layeredness does make things a bit more flexible and contained. For example with the graph, people can write their own grapher, and this can be then implemented without touching the html template. Though I'm guessing that people that would write a custom grapher would have no issue with editing an html template, it does raise the question how optional components would be integrated in the template system other than it is done now?

Working at the moment on a module system, basically a full refactor of the code where each module provides some output and takes some input, so that a user can exchange any part of obsidianhtml with his own implementation. 

That is to say, I don't foresee much time to work on restructuring the templating system atm, so I won't be putting time into this idea. BUT if the module system pans out, would you write code that uses your own templating system, and we can work together to get that put into a module that can exchange the then default templating module.

Something like that?

PR's are always welcome of course.
